lib/host_delete.py

```diff
@@ -1,41 +1,53 @@
+from kafka.errors import KafkaError
 from sqlalchemy.orm.base import instance_state
 
+from app.logging import get_logger
 from app.models import Host
 from app.queue.events import build_event
 from app.queue.events import EventType
 from app.queue.events import message_headers
+from lib.db import session_guard
+from lib.host_kafka import kafka_available
 from lib.metrics import delete_host_count
 from lib.metrics import delete_host_processing_time
 
 __all__ = ("delete_hosts",)
+logger = get_logger(__name__)
 
 
 def delete_hosts(select_query, event_producer, chunk_size, interrupt=lambda: False):
-    while select_query.count():
-        for host in select_query.limit(chunk_size):
-            host_id = host.id
-            with delete_host_processing_time.time():
-                _delete_host(select_query.session, host)
+    with session_guard(select_query.session):
+        while select_query.count():
+            for host in select_query.limit(chunk_size):
+                host_id = host.id
+                with delete_host_processing_time.time():
+                    host_deleted = _delete_host(select_query.session, event_producer, host)
 
-            host_deleted = _deleted_by_this_query(host)
-            if host_deleted:
-                delete_host_count.inc()
+                yield host_id, host_deleted
 
-                event = build_event(EventType.delete, host)
-                insights_id = host.canonical_facts.get("insights_id")
-                headers = message_headers(EventType.delete, insights_id)
-                event_producer.write_event(event, str(host.id), headers, wait=True)
+                if interrupt():
+                    return
 
-            yield host_id, host_deleted
 
-            if interrupt():
-                return
-
-
-def _delete_host(session, host):
+def _delete_host(session, event_producer, host):
     delete_query = session.query(Host).filter(Host.id == host.id)
-    delete_query.delete(synchronize_session="fetch")
-    delete_query.session.commit()
+    if kafka_available():
+        delete_query.delete(synchronize_session="fetch")
+        host_deleted = _deleted_by_this_query(host)
+        if host_deleted:
+            delete_host_count.inc()
+            event = build_event(EventType.delete, host)
+            insights_id = host.canonical_facts.get("insights_id")
+            headers = message_headers(EventType.delete, insights_id)
+            event_producer.write_event(event, str(host.id), headers, wait=True)
+            delete_query.session.commit()
+            return host_deleted
+        else:
+            delete_query.session.rollback()
+            return host_deleted
+    else:
+        logger.error(f"host with {host.id} NOT deleted because Kafka server not available.")
+        raise KafkaError("Kafka server not available.  Stopping host deletions.")
```

Comment:
I don't think that the code currently will stop _attempting_ to delete hosts when this error is raised. There is no exception handling done in the `delete_hosts` loop, so it will continue running and raising these exceptions.

I think this exception handling and structure need work. There's more nesting than needed, making it the responsibility of callers to raise the correct error when kafka can't be reached is error prone, and the exception you're raising is not being caught or acted upon.

I think you should instead check if Kafka is available first and if it's not re-raise the exception and return early.

```suggestion
    try:
        kafka_available()
    except KafkaError:
        logger.error(f"host with {host.id} NOT deleted because Kafka server not available.")
        raise

    delete_query.delete(synchronize_session="fetch")
    host_deleted = _deleted_by_this_query(host)
    if host_deleted:
        delete_host_count.inc()
        event = build_event(EventType.delete, host)
        insights_id = host.canonical_facts.get("insights_id")
        headers = message_headers(EventType.delete, insights_id)
        event_producer.write_event(event, str(host.id), headers, wait=True)
        delete_query.session.commit()
        return host_deleted
    else:
        delete_query.session.rollback()
        return host_deleted
```

Reply:
> I don't think that the code currently will stop _attempting_ to delete hosts when this error is raised. There is no exception handling done in the `delete_hosts` loop, so it will continue running and raising these exceptions.
> 
> I think this exception handling and structure need work. There's more nesting than needed, making it the responsibility of callers to raise the correct error when kafka can't be reached is error prone, and the exception you're raising is not being caught or acted upon.
> 
> I think you should instead check if Kafka is available first and if it's not re-raise the exception and return early.

The current state of implementation is the result of review comments from the team.  The first implementation was to check for Kafka availability before starting the deletion.  If Kafka was inaccessible, don't start deleting hosts.  Then Chris K pointed out what if Kakfa disappeared after some hosts got delete but not all, which was a valid point.

When the first KafkaError is raised for the lack of Kafka availability, exception is caught by the caller in "api.host.py::delete*"

Reply:
Gotcha, the raise stops the loop the function is called in. I thought you had to re-raise the exception for some reason.

I do think we could make this less nested though.
 
```suggestion
    if not kafka_available():
        logger.error(f"host with {host.id} NOT deleted because Kafka server not available.")
        raise KafkaError("Kafka server not available.  Stopping host deletions.")
    
    delete_query.delete(synchronize_session="fetch")
    host_deleted = _deleted_by_this_query(host)
    
    if not host_deleted:
        delete_query.session.rollback()
        return host_deleted

    delete_host_count.inc()
    event = build_event(EventType.delete, host)
    insights_id = host.canonical_facts.get("insights_id")
    headers = message_headers(EventType.delete, insights_id)
    event_producer.write_event(event, str(host.id), headers, wait=True)
    delete_query.session.commit()
    return host_deleted
```

Reply:
This suggestion has been implemented.